src/routes/owners/__tests__/get-safes-by-owner.e2e-spec.ts

```diff
@@ -0,0 +1,49 @@
+import { INestApplication } from '@nestjs/common';
+import { Test } from '@nestjs/testing';
+import { RedisClientType } from 'redis';
+import * as request from 'supertest';
+import { AppModule } from '../../../app.module';
+import { redisClientFactory } from '../../../__tests__/redis-client.factory';
+
+describe('Get safes by owner e2e test', () => {
+  let app: INestApplication;
+  let redisClient: RedisClientType;
+  const chainId = '5'; // Görli testnet
+
+  beforeAll(async () => {
+    const moduleRef = await Test.createTestingModule({
+      imports: [AppModule],
+    }).compile();
+
+    app = moduleRef.createNestApplication();
+    await app.init();
+    redisClient = await redisClientFactory();
+  });
+
+  beforeEach(async () => {
+    await redisClient.flushAll();
+  });
+
+  it('GET /owners/<owner_address>/safes', async () => {
```

Comment:
We do not cover situations where `TransactionApi. getSafesByOwner` fails for example.

Reply:
I generally agree, we should try to test as many scenarios as possible, but in this case mappings don't have any business logic (it's just one-to-one between domain and route interfaces) and regarding possible error scenarios I only see failures in connections to `TransactionApi/ConfigApi`. 

> We do not cover situations where `TransactionApi.getSafesByOwner` fails for example.

How do you think we could cover that? I mean, taking a look to `SafeRepository`, also `TransactionApiManager.getTransactionApi`/`ConfigApi.getChain` could fail, do you think we need to write a test case for every call? 

In `transaction-api.service.spec.ts` we have some test like `it('should forward error', () => {...})`, maybe we can write a test like that for every api call, but I think it could be a bit defensive, what do you think?

Reply:
> I generally agree, we should try to test as many scenarios as possible, but in this case mappings don't have any business logic (it's just one-to-one between domain and route interfaces) and regarding possible error scenarios I only see failures in connections to `TransactionApi/ConfigApi`.

In the test you are not completely testing the outcome:

```typescript
expect.objectContaining({ safes: expect.any(Array) })
```

If the CGW maps a `String` to an integer this would still be successful because it's an array.

> > We do not cover situations where `TransactionApi.getSafesByOwner` fails for example.
> 
> How do you think we could cover that? I mean, taking a look to `SafeRepository`, also `TransactionApiManager.getTransactionApi`/`ConfigApi.getChain` could fail, do you think we need to write a test case for every call?

We have all the testing setup to test this. Please see the reference in other tests. Testing every combination can be cumbersome and I'm not asking that but we are not testing any error scenario at all.

> In `transaction-api.service.spec.ts` we have some test like `it('should forward error', () => {...})`, maybe we can write a test like that for every api call, but I think it could be a bit defensive, what do you think?

That's not the goal. But What if the specification is to that if the transactions service fails we return a default payload and not forward the error? With the tests as they are I can change the error to return a default implementation and it would still work even though it's not correct.

Reply:
Regarding the 
```javascript
expect.objectContaining({ safes: expect.any(Array) })
```
we could test every element in the array is a string, but I think you're aiming for changing the testing approach here and doing a unit/controller test instead of a e2e, to allow us to control both the response and the errors emitted by the external dependencies, am I right? 🙂

Reply:
> Regarding the
> 
> ```js
> expect.objectContaining({ safes: expect.any(Array) })
> ```
> 
> we could test every element in the array is a string, but I think you're aiming for changing the testing approach here and doing a unit/controller test instead of an e2e, to allow us to control both the response and the errors emitted by the external dependencies, am I right? 🙂

I don't have a strong opinion if it should be an E2E or Unit test. I think having E2E can be more tricky to test some conditions.

Example:

1) Safe Transaction Service returns: `safe_address_1, safe_address_2, safe_address_3`
2) Nest Client "maps" and returns those

a. You can think of a test which just tests that it returns an array – testing the "schema" (although not the types of the schema).

The problem with this is that if the CGW returns an empty array the test would be successful.

b. To improve on a. you can write a test that tests the number of items of the array and if they are all of a certain type.

The problem with this is that we can have a bug where the number of items is the same and they are of the type specified and the test would not catch this.

c. To improve on b. we can test that the output is `safe_address_1, safe_address_2, safe_address_3`

With this we accomplish one additional thing: the order of the response returned – this might have a meaning on the client side.

Therefore even if the test that you have works to validate that it is an array, if a bug is introduced on this service the test would still be successful! In those cases the test should fail because the spec deviated from what is expected.

----

This might be easier to accomplish with a Unit test than E2E test (if the data is not "stable" on the services that we are testing). But my point is not about E2E test vs Unit test but about what are we testing.

Tests should help us detect potential underlying issues. A bug that we introduce should raise a warning before it's merged.

Reply:
Agree, there are some scenarios where E2E es more effective (_static_ blockchain data mostly), and there are others where a unit test is preferable. In this case, I think the data is mutable (the safes of a given owner can change over time) so maybe a unit test with a fixed scenario is better. Having that said, I think maybe we can combine both approaches here a test just a happy path and the Redis cache in the E2E, and add some unit tests to check the mappings, errors, and business logic in general, what do you think?

Reply:
I included some additional unit tests in [470799d](https://github.com/5afe/safe-client-gateway-nest/pull/192/commits/470799dcc5e257115ca6e95c6dbc7b2280abafc5), please let me know if you think something is missing.